users/user-guide-query/multi-stage-query/operator-types/hash_join.md

```diff
@@ -0,0 +1,210 @@
+---
+description: >-
+  Describes the hash join relation operator in the multi-stage query engine.
+---
+
+# Hash join relational operator
+
+The hash join operator is used to join two relations using a hash join algorithm. 
+It is a binary operator that takes two inputs, the left and right relations, and produces a single output relation. 
+
+This is the only join operator in the multi-stage query engine and it is always created as a result of a query that 
+contains a join clause, but can be created by other SQL queries like ones using semi-join.
+
+There are different types of joins that can be performed using the hash join operator.
+Apache Pinot supports:
+- Inner join, where only the rows that have a match in both relations are returned.
+- Left join, where all the rows from the left relation are returned. The ones that have a match with the right relation 
+  are returned with the columns from the right relation, and the ones that do not have a match are returned with null 
+  values for the columns from the right relation.
+- Right join, like the left join but returning all the rows from the right relation, with the columns from the left 
+  relation filled with null values for the rows that do not have a match.
+- Full outer join, where all the rows from both relations are returned. If a row from any relation does not have a match 
+  in the other relation, the columns from the other relation are filled with null values.
+- Semi-join, where only the rows from the left relation that have a match in the right relation are returned. This is 
+  useful to filter the rows from the left relation based on the existence of a match in the right relation.
+- Anti-join, where only the rows from the left relation that do not have a match in the right relation are returned.
+
+## Implementation details
+
+The hash join operator is one of the new operators introduced in the multi-stage query engine.
```

Comment:
noob question - in shuffle hash, how does it matter that right is smaller, if both tables are getting hashed and then sent to a new server? is it because some groups may not exist at all, so we have a chance of skipping sending something over from left to the shuffle node?

Reply:
Only the right hand side is being used to create the hash table. That hash table is calculated in every worker and depending on the distribution used it may be calculated using all the values from the right hand side (in case of broadcasting) or only a partition.

Once the hash table is built, then the worker starts to consume from the left hand side. For each row it probes whether there is a hit in the hash table or not. If the row should be kept (ie we are executing a inner join) then a new row is added to the block to be emited.

Therefore:
- Rows in the right hand side are used to create an in memory structure that is kept in memory as long as the operator is running.
- Rows in the left hand side are compared with the hash table and optionally add one or more rows on the block to be emitted.

It is also important to understand that each input can produce more than one data block and for each block received from the left hand side, the operator will emit a new block (assuming not all values have been filtered out) for each block received from the left hand side.

Reply:
`Once the hash table is built, then the worker starts to consume from the left hand side. ` - but this would mean we're always executing as broadcast, unless data is pre-partitioned?

Reply:
Not always. I think that is explained somewhere else, but it is probably worth it to have a specific page to discuss about that.

We can say that by default we apply broadcast. But there are several optimizations we apply for free. For example in a query like `select ... from A join B on A.col1 = A.col2` we can apply hashing on A.col1 and B.col2. Imagine we have 10 workers, then each worker will receive a partition of the data (basically all rows whose `hash(A.col1) = k` or `hash(B.col2) = k` where `k` is unique for each worker). This is the most common case.

But in queries where we cannot apply partition (ie `select ... from A join B on A.col1 < B.col2` will apply partition on the left hand side (A in this case) but we need to broadcast the right hand side (B in this case). In this cases is it even more important to keep the left hand side as the smallest relation, given an entire copy will be sent to each worker.